src/java/org/apache/cassandra/service/accord/serializers/EncodeAsVInt32.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.service.accord.serializers;
+
+import java.io.IOException;
+import java.util.function.IntFunction;
+import java.util.function.ToIntFunction;
+
+import javax.annotation.Nullable;
+
+import com.google.common.primitives.Ints;
+
+import org.apache.cassandra.io.IVersionedSerializer;
+import org.apache.cassandra.io.util.DataInputPlus;
+import org.apache.cassandra.io.util.DataOutputPlus;
+import org.apache.cassandra.utils.vint.VIntCoding;
+
+public abstract class EncodeAsVInt32<T> implements IVersionedSerializer<T>
+{
+    public static <T> EncodeAsVInt32<T> withNulls(ToIntFunction<? super T> encode, IntFunction<? extends T> decode)
+    {
+        return new WithNulls<>(encode, decode);
+    }
+
+    public static <T> EncodeAsVInt32<T> withoutNulls(ToIntFunction<? super T> encode, IntFunction<? extends T> decode)
+    {
+        return new WithoutNulls<>(encode, decode);
+    }
+
+    public static <E extends Enum<?>> EncodeAsVInt32<E> of(Class<E> clazz)
+    {
+        E[] values = clazz.getEnumConstants();
+        return withNulls(Enum::ordinal, i -> values[i]);
```

Comment:
given our project has a stance of `@Nonnull` is the default and you only document when `@Nullable` is allowed (as its not the default), then shouldn't this use `withoutNulls`?

Reply:
The serialiser's capability is distinct from annotation expectations. There's zero reason to have separate serializers for nullable and not-nullable. I'm not even sure it makes sense to support a `withoutNulls` option here, really.

Reply:
I guess zero reason is a bit strong - there's a nominal additional capability to detect corruption. But it's so close to no effect (rejecting precisely one additional integer value in that stream location), I do not think it is likely to ever be worth the complexity of additional serializer instances.